### PR TITLE
exercise 4.1

### PR DESCRIPTION
**Pull request by Lira999**

### CODE REVIEW

File: exercise_4/src/CommandHandler.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+#include <iostream>
+#include <string>
+#include <map>
+#include <memory>
+
+#include "ICommand.h"
+#include "IRecord.h"
+
+class CommandHandler : public ICommand {
```
Comment (mjtieto) [inline exercise_4/src/CommandHandler.h:11]: I think `CommandHandler` is not a good approach for commands execution.
Every command (Show, Delete, Update) should be implemented as a separate class, that derives from the ICommand interface.
Thanks to that, we don't introduce a god object (CommandHandler) that knows about all the commands, but we have a couple of simple classes, that don't know (and don't need to know) about any other command.

Imagine a scenario, where you have not three, but 300 possible different commands.
If you'd want to handle 300 different commands in the `CommandHandler`, you'd end up with a huge, unmaintainable class.

File: exercise_4/src/CommandHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include "CommandHandler.h"
+
+CommandHandler::CommandHandler(const std::map<std::string, std::map<int, std::shared_ptr<IRecord> > > &records,
+	                           const std::string &command,
+	                           const std::string &table_name): _records(records),
+                                                               _command(command),
+                                                               _table_name(table_name){}
+
+void CommandHandler::execute() {
+	std::cout << _command << _table_name << std::endl;
```
Comment (mjtieto) [inline exercise_4/src/CommandHandler.cpp:15]: Such debug prints should not be pushed to a 'production' code.

File: exercise_4/src/CommandHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		_show();
+	} else {
+		std::cerr << "Unknown command." << std::endl;
+	}
+
+}
+
+void CommandHandler::_show() {
+	if (_table_name == "") {
+		std::cout << "Pretty prints whole database to the output. Records are grouped into tables:" << std::endl;
```
Comment (mjtieto) [inline exercise_4/src/CommandHandler.cpp:26]: I don't think this print is necessary

File: exercise_4/src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+#include <memory>
+
+#include "IRecord.h"
+#include "Course.h"
+#include "Exam.h"
+#include "Student.h"
+#include "Teacher.h"
+#include "CommandHandler.h"
+
+using namespace std;
```
Comment (mjtieto) [inline exercise_4/src/main.cpp]: Don't use `using namespace ...`, especially the `std` one.
https://stackoverflow.com/questions/1452721/why-is-using-namespace-std-considered-bad-practice

File: exercise_4/src/main.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+int main(int args, char **argv) {
+	if (args > 1) {
+		ifstream input_file("./src/records.txt");
+		string line, marker, command_type, table_name = "";
+
+		map<int, shared_ptr<IRecord> > teachers;
+		map<int, shared_ptr<IRecord> > students;
+		map<int, shared_ptr<IRecord> > courses;
+		map<int, shared_ptr<IRecord> > exams;
```
Comment (mjtieto) [inline exercise_4/src/main.cpp]: You don't need to share ownership, so `shared_ptr` is not needed => use `unique_ptr`. Same in the `container`

**Review by mjtieto — CHANGES_REQUESTED** [review]

File: exercise_4/src/CommandHandler.h (hunk 1 above)
Comment (Lira999) [reply exercise_4/src/CommandHandler.h:11]: Renamed.

File: exercise_4/src/CommandHandler.cpp (hunk 2 above)
Comment (Lira999) [reply exercise_4/src/CommandHandler.cpp:15]: Removed print.

File: exercise_4/src/CommandHandler.cpp (hunk 3 above)
Comment (Lira999) [reply exercise_4/src/CommandHandler.cpp:26]: Removed print.

File: exercise_4/src/main.cpp (hunk 4 above)
Comment (Lira999) [reply exercise_4/src/main.cpp]: Removed.

File: exercise_4/src/main.cpp (hunk 5 above)
Comment (Lira999) [reply exercise_4/src/main.cpp]: Done.

**Pull request merged**